tests/unit/test_coordinator.py

```diff
@@ -587,6 +594,111 @@ def test_invalid_app_or_unit_databag(
+        assert len(job["static_configs"]) == 1
+        assert "targets" in job["static_configs"][0]
+        assert "labels" in job["static_configs"][0]
+        assert "juju_unit" in job["static_configs"][0]["labels"]
+
+
+def test_coordinator_scrape_jobs_generation_single_unit(
+    coordinator_charm: ops.CharmBase,
+    nginx_container,
+    nginx_prometheus_exporter_container,
+    mock_coordinator_get_peer_data_patch,
+):
+    ctx = testing.Context(coordinator_charm, meta=coordinator_charm.META, app_name="coordinator")
+    my_metrics_relation = testing.Relation(endpoint="my-metrics")
+
+    # SCENARIO 1: Single unit deployment
+    # GIVEN: A coordinator with a single unit (no remote peers)
+    peer_relation = testing.PeerRelation(endpoint="my-peers")
+    single_unit_state = testing.State(
+        leader=True,
+        relations=[peer_relation, my_metrics_relation],
+        containers={
+            nginx_container,
+            nginx_prometheus_exporter_container,
+        },
+    )
+
+    # WHEN: No remote peers are present (only local unit)
+    single_remote_peers = {}  # No remote peers
+    with mock_coordinator_get_peer_data_patch(single_remote_peers):
+        with ctx(ctx.on.update_status(), single_unit_state) as manager:
+            manager.run()
+            single_jobs = manager.charm.coordinator._scrape_jobs
+            _validate_scrape_jobs(single_jobs, 1)
+
+    # THEN: Exactly one scrape job should be generated for the local unit
+    assert single_jobs[0]["static_configs"][0]["labels"]["juju_unit"] == "coordinator/0"
+    # And: The target should use the correct port format (hostname:9113)
+    assert single_jobs[0]["static_configs"][0]["targets"][0].endswith(":9113")
+
+
+def test_coordinator_scrape_jobs_generation_scaled(
+    coordinator_charm: ops.CharmBase,
+    nginx_container,
+    nginx_prometheus_exporter_container,
+    mock_coordinator_get_peer_data_patch,
+):
+    ctx = testing.Context(coordinator_charm, meta=coordinator_charm.META, app_name="coordinator")
+    my_metrics_relation = testing.Relation(endpoint="my-metrics")
+    # GIVEN: A coordinator with 3 units (local + 2 remote peers)
+    scaled_peer_relation = testing.PeerRelation(endpoint="my-peers")
+    scaled_state = testing.State(
+        leader=True,
+        relations=[scaled_peer_relation, my_metrics_relation],
+        containers={
+            nginx_container,
+            nginx_prometheus_exporter_container,
+        },
+    )
+
+    # WHEN: Two remote peers are available with their hostnames
+    scaled_remote_peers = {}
+    remote_unit_names = ["coordinator/1", "coordinator/2"]
+
+    for unit_name in remote_unit_names:
+        mock_unit = Mock()
+        mock_unit.name = unit_name
+        # Use simple hostname matching unit name pattern
+        hostname = unit_name.replace("coordinator/", "coord-") + ".local"
+        scaled_remote_peers[mock_unit] = hostname
+
+    with mock_coordinator_get_peer_data_patch(scaled_remote_peers):
```

Comment:
this is how I'd get rid of the patch: https://github.com/canonical/cos-coordinated-workers/pull/110